python/pyiceberg/io/pyarrow.py

```diff
@@ -517,15 +536,22 @@ def _file_to_table(
         if file_schema is None:
             raise ValueError(f"Missing Iceberg schema in Metadata for file: {path}")
 
-        arrow_table = pq.read_table(
-            source=fout,
-            schema=parquet_schema,
-            pre_buffer=True,
-            buffer_size=8 * ONE_MEGABYTE,
-            filters=pyarrow_filter,
+        fragment_scanner = ds.Scanner.from_fragment(
+            fragment=fragment,
+            schema=physical_schema,
+            filter=pyarrow_filter,
             columns=[col.name for col in file_project_schema.columns],
         )
 
+        if limit:
+            arrow_table = fragment_scanner.head(limit)
+            with rows_counter.get_lock():
```

Comment:
I think we can remove this lock because we already did the expensive work. This will make the code a bit simpler and avoid locking.

Reply:
The [official python documentation for multiprocessing.Value](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Value) suggests that non atomic operations like `+=` should use a lock. Otherwise a race condition could happen where multiple reads end at the same time and overwrite the row counter, causing that we keep reading rows even if we already read enough

Reply:
Ah, thanks for the explanation. Currently, we don't do multi-processing, but multi-threading. I did some extensive testing and noticed that multi-processing wasn't substantially faster than multithreading. Probably because most time is spent in fetching the files, and reading the data, which all happens in Arrow, which bypasses the GIL.

Reply:
I think you actually made the right choice going for multi-threading because the task is [IO-bound and there are potentially thousands of tasks that need to be executed](https://superfastpython.com/threadpool-vs-pool-in-python/). Even though we use multi-threading which is bound to the GIL [some operations might still not be safe](https://superfastpython.com/threadpoolexecutor-thread-safe/) (like `+=` which performs a read and a write as two separate operations).
Also, even though the tasks are run on threads in a `ThreadPool`, [`multiprocessing.Value`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Value) implements a [`multiprocessing.RLock`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.RLock) by default wich is compatible with both processes and threads.

Reply:
Thanks for clearing this up!